### PR TITLE
Improve conflict marker detection and fixing

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/23847

Implementation is revised somewhat.  Instead of being based on trivia (which may not exist in the case of comments/strings), it is now based on trivia+lines

### CODE REVIEW

File: src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs [hunk 1]
```diff
@@ -148,6 +228,7 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
                 context.Diagnostics);
         }
 
+
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs]:
```suggestion
```

File: src/EditorFeatures/CSharpTest/ConflictMarkerResolution/ConflictMarkerResolutionTests.cs [hunk 2]
```diff
@@ -529,6 +529,154 @@ static void Main2(string[] args)
...
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsResolveConflictMarker)]
+        public async Task TestTakeTop_TopCommentedOut()
+        {
+            var source = @"
+public class Class1
+{
+    public void M()
+    {
+        /*
+<<<<<<< dest
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/ConflictMarkerResolution/ConflictMarkerResolutionTests.cs]: 💭 I guess we'll need an analyzer to cover the case of markers entirely inside comments.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/ConflictMarkerResolution/ConflictMarkerResolutionTests.cs]: Yup.  But we can wait on that for now IMO.

**Comment by CyrusNajmabadi** [comment]
@sharwell good to proceed on this PR?

**Comment by sharwell** [comment]
@CyrusNajmabadi I only reviewed the tests

**Comment by JoeRobich** [comment]
I use VS Code as my mergetool and have become a fan of the diff3 conflict style (https://stackoverflow.com/questions/27417656/should-diff3-be-default-conflictstyle-on-git), since it allows me to resolve issues with more confidence. Can we add a test showing that we do the right thing when encountering these style of markings?

```
<<<<<<< HEAD
THIS IS USEFUL
||||||| merged common ancestors
This is useful
=======
This is really useful
>>>>>>> c2392943.....
````

**Comment by CyrusNajmabadi** [comment]
@JoeRobich we would not do the right hting there (we don't recognize `|||||||` in the lexer as being special).  I would be totally fine adding support for that (and would be happy to walk you through doing this), but it should be in anotehr issue/pr, thanks!

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide @mavasani ptal.

File: src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs [hunk 3]
```diff
@@ -45,91 +45,171 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
...
 
-            var startTrivia = root.FindTrivia(context.Span.Start);
-            if (!IsConflictMarker(text, startTrivia, '<'))
+            var position = context.Span.Start;
+            var conflictLine = lines.GetLineFromPosition(position);
+            if (position != conflictLine.Start)
+            {
+                Debug.Assert(false, "All conflict markers should start at the beginning of a line.");
                 return;
+            }
```
Comment (genlu) [inline src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs]: This check is redundant, is part of `ShouldFix` too.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs]: true!

File: src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs [hunk 4]
```diff
@@ -45,91 +45,171 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
...
+                    foundLine = currentLine;
+                    return true;
                 }
+            }
+
+            foundLine = default;
+            return false;
+        }
+
+        private bool IsConflictMarker(TextLine currentLine, char ch)
```
Comment (genlu) [inline src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs]: nit: many members can be static

File: src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs [hunk 5]
```diff
@@ -45,91 +45,171 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
...
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(text[position]);
+            }
+        }
+
+        private bool TryFindLineForwards(TextLine startLine, char ch, out TextLine foundLine)
+        {
+            var text = startLine.Text!;
+            var lines = text.Lines;
+            for (var i = startLine.LineNumber + 1; i < lines.Count; i++)
```
Comment (genlu) [inline src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs]: startLine might be the last line

File: src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs [hunk 6]
```diff
@@ -45,91 +45,171 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
...
-            while (true)
+            foundLine = default;
+            return false;
+        }
+
+        private bool TryFindLineBackwards(TextLine startLine, char ch, out TextLine foundLine)
+        {
+            var text = startLine.Text!;
+            var lines = text.Lines;
+            for (var i = startLine.LineNumber - 1; i >= 0; i--)
```
Comment (genlu) [inline src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs]: Similarly, startLine can be the first line too.

File: src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs]: done!

File: src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs]: i'm not sure what this means.

File: src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs]: note: in that case, this for-loop will bail out immediately.

File: src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs]: in that case, the for loop will bail out immediatley.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs]: i will add tests :)

File: src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs (hunk 5 above)
Comment (genlu) [reply src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs]: You are right. I was a little confused why there's a diagnostic even if not all markers exist. So, is the diagnostic created by compiler?

---
In reply to: [395393643](https://github.com/dotnet/roslyn/pull/42213#discussion_r395393643) [](ancestors = 395393643)

File: src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs]: added.

File: src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs]: added test.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs]: > You are right. I was a little confused why there's a diagnostic even if not all markers exist. So, is the diagnostic created by compiler?

Yes.  diagnsotic is created by compiler when tehy see `<<<<<<` and the like.  it becomes trivia (normally) and they issue a diagnostic.

This worked well initially for the codefix, but falls over in the case where the stuff is in string-literals.  So i'm fixing that up now.  We still need at least one of these to be in real code.  but as long as tehre's one, we can fix them up :)

File: src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs [hunk 7]
```diff
@@ -45,91 +45,163 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
...
                 }
+            }
 
-                token = token.GetNextToken(includeZeroWidth: true);
-                if (token.RawKind == 0)
+            foundLine = default;
+            return false;
+        }
+
+        private static bool IsConflictMarker(TextLine currentLine, char ch)
```
Comment (genlu) [inline src/Features/Core/Portable/ConflictMarkerResolution/AbstractConflictMarkerCodeFixProvider.cs]: So this means we don't handle this properly, right? I think it's not unreasonable, just want to confirm

image

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**